src/test/func/malloc/malloc.cc

```diff
@@ -0,0 +1,118 @@
+#include <cerrno>
+#include <malloc.h>
+#include <snmalloc.h>
+
+using namespace snmalloc;
+
+void check_result(size_t size, size_t align, void* p, int err, bool null)
+{
+  if (errno != err)
+    abort();
+
+  if (null)
+  {
+    if (p != nullptr)
+      abort();
+  }
+  else
+  {
+    if (malloc_usable_size(p) < size)
+      abort();
+
+    if (((uintptr_t)p % align) != 0)
+      abort();
+
+    free(p);
+  }
+}
+
+void test_calloc(size_t nmemb, size_t size, int err, bool null)
+{
+  fprintf(stderr, "calloc(%d, %d)\n", (int)nmemb, (int)size);
+  errno = 0;
+  void* p = calloc(nmemb, size);
+
+  if ((p != nullptr) && (errno == 0))
+  {
+    for (size_t i = 0; i < (size * nmemb); i++)
+    {
+      if (((uint8_t*)p)[i] != 0)
+        abort();
+    }
+  }
+  check_result(nmemb * size, 1, p, err, null);
+}
+
+void test_realloc(void* p, size_t size, int err, bool null)
+{
+  fprintf(stderr, "realloc(%p(%d), %d)\n", p, int(size), (int)size);
+  errno = 0;
+  p = realloc(p, size);
+  check_result(size, 1, p, err, null);
+}
+
+void test_posix_memalign(size_t size, size_t align, int err, bool null)
+{
+  fprintf(stderr, "posix_memalign(&p, %d, %d)\n", (int)align, (int)size);
+  void* p = nullptr;
+  errno = posix_memalign(&p, align, size);
+  check_result(size, align, p, err, null);
+}
+
+void test_memalign(size_t size, size_t align, int err, bool null)
+{
+  fprintf(stderr, "memalign(%d, %d)\n", (int)align, (int)size);
+  errno = 0;
+  void* p = memalign(align, size);
```

Comment:
memalign is deprecated in glibc and doesn't exist anywhere else, so we probably need to explicitly declare it if we want to test our version.